lib/evmone/instructions.hpp

```diff
@@ -104,1170 +103,1382 @@ inline constexpr int64_t copy_cost(uint64_t size_in_bytes) noexcept
+    tbl[EVMC_PARIS] = tbl[EVMC_LONDON];
+    tbl[EVMC_SHANGHAI] = tbl[EVMC_LONDON];
+    tbl[EVMC_CANCUN] = tbl[EVMC_LONDON];
+    tbl[EVMC_PRAGUE] = tbl[EVMC_LONDON];
+    return tbl;
+}();
+
+
+struct StorageStoreCost
+{
+    int16_t gas_cost;
+    int16_t gas_refund;
+};
 
-/// Check memory requirements for "copy" instructions.
-inline bool check_memory(
-    int64_t& gas_left, Memory& memory, const uint256& offset, const uint256& size) noexcept
-{
-    if (size == 0)  // Copy of size 0 is always valid (even if offset is huge).
-        return true;
+// The lookup table of SSTORE costs by the storage update status.
+constexpr auto sstore_costs = []() noexcept {
+    std::array<std::array<StorageStoreCost, EVMC_STORAGE_MODIFIED_RESTORED + 1>,
+        EVMC_MAX_REVISION + 1>
+        tbl{};
 
-    // This check has 3 same word checks with the check above.
-    // However, compilers do decent although not perfect job unifying common instructions.
-    // TODO: This should be done in intx.
-    if (((size[3] | size[2] | size[1]) != 0) || (size[0] > max_buffer_size))
-        return false;
+    for (size_t rev = EVMC_FRONTIER; rev <= EVMC_MAX_REVISION; ++rev)
+    {
+        auto& e = tbl[rev];
+        if (const auto c = storage_cost_spec[rev]; !c.net_cost)  // legacy
+        {
+            e[EVMC_STORAGE_ADDED] = {c.set, 0};
+            e[EVMC_STORAGE_DELETED] = {c.reset, c.clear};
+            e[EVMC_STORAGE_MODIFIED] = {c.reset, 0};
+            e[EVMC_STORAGE_ASSIGNED] = e[EVMC_STORAGE_MODIFIED];
+            e[EVMC_STORAGE_DELETED_ADDED] = e[EVMC_STORAGE_ADDED];
+            e[EVMC_STORAGE_MODIFIED_DELETED] = e[EVMC_STORAGE_DELETED];
+            e[EVMC_STORAGE_DELETED_RESTORED] = e[EVMC_STORAGE_ADDED];
+            e[EVMC_STORAGE_ADDED_DELETED] = e[EVMC_STORAGE_DELETED];
+            e[EVMC_STORAGE_MODIFIED_RESTORED] = e[EVMC_STORAGE_MODIFIED];
+        }
+        else  // net cost
+        {
+            e[EVMC_STORAGE_ASSIGNED] = {c.warm_access, 0};
+            e[EVMC_STORAGE_ADDED] = {c.set, 0};
+            e[EVMC_STORAGE_DELETED] = {c.reset, c.clear};
+            e[EVMC_STORAGE_MODIFIED] = {c.reset, 0};
+            e[EVMC_STORAGE_DELETED_ADDED] = {c.warm_access, static_cast<int16_t>(-c.clear)};
+            e[EVMC_STORAGE_MODIFIED_DELETED] = {c.warm_access, c.clear};
+            e[EVMC_STORAGE_DELETED_RESTORED] = {
+                c.warm_access, static_cast<int16_t>(c.reset - c.warm_access - c.clear)};
+            e[EVMC_STORAGE_ADDED_DELETED] = {
+                c.warm_access, static_cast<int16_t>(c.set - c.warm_access)};
+            e[EVMC_STORAGE_MODIFIED_RESTORED] = {
+                c.warm_access, static_cast<int16_t>(c.reset - c.warm_access)};
+        }
+    }
 
-    return check_memory(gas_left, memory, offset, static_cast<uint64_t>(size));
-}
+    return tbl;
+}();
 
 namespace instr::core
 {
+template <typename BlueprintFieldType>
+struct operation {
```

Comment:
Maybe `instruction/instructions`?

Reply:
do you mean rename struct to `instructions`?

Reply:
Yes, `operation` is unclear, I think it's better to align the name of the struct with the name of the file.